ceph/volume.go

```diff
@@ -0,0 +1,190 @@
+}
+
+const (
+	CephImageSizeMB   = 512 // 1TB
+	LuksDevMapperPath = "/dev/mapper/"
+	cryptoLuksFsType  = "crypto_LUKS"
+)
+
+func (v *Volume) Info() map[string]string {
+	return map[string]string{
+		OPT_VOLUME_NAME: v.Name,
+		OPT_MOUNT_POINT: v.MountPoint,
+		"Device":        v.Device,
+		"LuksDevice":    v.LUKSDevice,
+	}
+}
+
+func (v *Volume) mapCephVolume() error {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.Command("rbd", "map", v.Name)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	var Device string
+	if err := cmd.Run(); err == nil {
+		Device = strings.TrimRight(stdout.String(), "\n")
+		log.Infof("Succeeded in mapping Ceph volume='%s' to device=%s", v.Name, Device)
+		v.Device = Device
+		return nil
+	} else {
+		msg := fmt.Sprintf("Failed to map Ceph volume='%s': error=%s - stderr=%s", v.Name, err, strings.TrimRight(stderr.String(), "\n"))
+		log.Errorf(msg)
+		return errors.New(msg)
+	}
+}
+
+func (v *Volume) unmapCephVolume() error {
+	cmd := exec.Command("rbd", "unmap", v.Device)
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err == nil {
+		log.Infof("Succeeded in unmapping Ceph volume='%s' from device=%s", v.Name, v.Device)
+		v.Device = ""
+	} else {
+		log.Errorf("Failed to unmap Ceph volume='%s' from device=%s: error=%s - stderr=%s", v.Name, v.Device, err, strings.TrimRight(stderr.String(), "\n"))
+	}
+	return err
+}
+
+func (v *Volume) Map(id string, sizeMB int64) (Device string, returnedError error) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	// TODO: Might want to map with --options rw/ro here, but then we need to sneak in the RW flag somehow
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command("rbd", "create", v.Name, "--size", fmt.Sprintf("%v", sizeMB))
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err == nil {
+		log.Infof("Created Ceph volume '%s'", v.Name)
+	} else {
+		// if rbd create returned EEXIST (17) the image is already there and we just need to map
+		if exitError, ok := err.(*exec.ExitError); ok {
+			imageSpec := strings.Split(v.Name, "/") // strip the pool from the name
+			imageName := imageSpec[len(imageSpec)-1]
+			waitStatus := exitError.Sys().(syscall.WaitStatus)
+			if waitStatus.ExitStatus() == 17 || strings.Contains(stderr.String(), fmt.Sprintf("rbd image %s already exists", imageName)) {
+				log.Infof("Found existing Ceph volume='%s' - error=%s", v.Name, err)
+			} else {
+				msg := fmt.Sprintf("Failed to create Ceph volume='%s'. stderr=%s. exitcode=(%d) ", v.Name, stderr.String(), waitStatus.ExitStatus())
+				log.Errorf(msg)
+				return "", errors.New(msg)
+			}
+		} else {
+			log.Errorf(fmt.Sprintf("Failed to get exit code from ceph volume creation (volume='%s') - error=%s", v.Name, stderr.String()))
+			return "", err
+		}
+	}
+	if err := v.mapCephVolume(); err != nil {
+		return "", err
+	}
+	return v.Device, nil
+}
+
+func (v *Volume) LUKSEncryption() error {
```

Comment:
LUKSEncryption should be LUKSOpen since it is actually performing an action to open the encrypted device